### PR TITLE
WIP: Compiler Plugins

### PR DESCRIPTION
**Pull request by rktjmp**
Breaking Changes:

- ~~`exclude_keys` removed in favor of `vim_compatible` plugin.~~

Potential Breaking Changes:

- Nest `lush(..., options)` as `options = {compiler: ..., parser: ...}`, which would break the current `force_clean` option (it would end up as `{compiler: {force_clean: true}}`. 
- `lush.export_to_buffer`, superseded

### CODE REVIEW

File: lua/lush/compiler.lua [hunk 1]
```diff
@@ -1,87 +1,48 @@
...
+  end
+
   for group_name, group_def in pairs(ast) do
-    local command
+    local command = ""
+    local maybe_halt = nil
+
     if group_def.link then
-      command = make_link(group_name, group_def.link, options)
+      for _, plug in ipairs(plugins) do
```
Comment (rktjmp) [inline lua/lush/compiler.lua:29]: outer wrap

File: lua/lush/compiler/plugin/vim_compatible.lua [hunk 2]
```diff
@@ -0,0 +1,22 @@
+local value_or_NONE = require("lush.compiler.plugin.utils").value_or_NONE
+
+local vim_compatible = {
+  name = "lush vim-compatible",
+  make_group = function(group_name, group_table, current_rule, entire_spec)
+    -- -- vim cant handle blend values
```
Comment (rktjmp) [inline lua/lush/compiler/plugin/vim_compatible.lua:6]: double markers

File: lua/lush/compiler/plugin/vim_compatible.lua [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    local modified_rule
+    modified_rule = string.gsub(current_rule, "blend=NONE", "")
+    modified_rule = string.gsub(modified_rule, "blend=%d+", "")
+
+    -- -- vim also handles gui values differently
```
Comment (rktjmp) [inline lua/lush/compiler/plugin/vim_compatible.lua:11]: double markers

File: lua/lush/exporter.lua [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  -- pass through the pipeline
+  for i, transform in ipairs(pipeline) do
+    if type(transform) == "function" then
+      -- raw function, just value -> value
+      value = transform(value)
+    elseif type(transform) == "table" then
+      -- table, first element must be the transformer, the rest are assumed to
+      -- be arguments for the transformer, excepting that the *first* argument
+      -- should be the current value.
+      assert(#transform > 0, lush_module .. " transformation # " .. i .. " was table with length 0")
```
Comment (rktjmp) [inline lua/lush/exporter.lua:28]: linebreak

File: lua/lush/exporter.lua [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  -- It's expected that one of the transformations has done something
+  -- useful with the work in terms of writing to disk or loading into memory.
+  return value
+end
+
+-- Create an environment to run the build file in. This should expose all the built in
+-- transformers, as well as lush itself.
+local function make_env()
+  local env = {
+    vim = vim,
```
Comment (rktjmp) [inline lua/lush/exporter.lua:50]: can remove due to _G proxy

File: lua/lush/transformer/patch.lua [hunk 6]
```diff
@@ -0,0 +1,8 @@
+-- Tail exporter, accepts a table of strings and a path
```
Comment (rktjmp) [inline lua/lush/transformer/patch.lua:1]: Possibly term "patchwrite" to pair with "overwrite"

File: spec/build_spec.moon [hunk 7]
```diff
@@ -0,0 +1,45 @@
+describe "lush", ->
+  lush = require('lush')
+  nvim_command_spy, red, green, blue = nil
+  lush_spec = nil
+
+  before_each ->
+    vim = {
```
Comment (rktjmp) [inline spec/build_spec.moon:7]: Can slim this mock?

**Review by rktjmp — COMMENTED** [review]
:eye: :lips: :eye:

File: lua/lush/compiler.lua [hunk 8]
```diff
@@ -1,87 +1,48 @@
...
-    if v == value then return true end
-  end
-  return false
-end
-
 -- make sure options is usable
 local function normalise_options(options)
   if options == nil then options = {} end
-  if options.exclude_keys == nil then options.exclude_keys = {} end
+  if options.plugins == nil then options.plugins = {} end
```
[the remark this answers is not in the document]
Comment (rktjmp) [reply lua/lush/compiler.lua:4]: No. We wont soft deprecate it, so the warning isn't much different to it just crashing.

File: BUILD.md [hunk 9]
```diff
@@ -37,9 +37,9 @@ Transforms can take any additional number of arguments after the table.
 Lush ships with some default transforms, which are automatically injected into
 the build environment:
 
-- `viml`, head, convert parsed spec into viml
-- `lua`, head, convert parsed spec into lua
-- `vim_compatible`, tail, remove vim-incompatible values from highlight rules
+- `lush_to_viml`, head, convert parsed spec into viml
```
Comment (mcchrish) [inline BUILD.md:40]: I think people are moving away from viml term in favor of "vim script". https://github.com/neovim/neovim/issues/16150

Comment (rktjmp) [reply BUILD.md:40]: I have *always* hated "viml", feels like I am throwing up in my mouth but was never sure which to use.

:+1:

**Pull request merged**